iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp

```diff
@@ -234,9 +234,15 @@ static LogicalResult buildModuleDescriptors(IREE::VM::ModuleOp &moduleOp,
   // module descriptor
   // TODO(simon-camp): support module-level reflection attributes
   std::string descriptorName = moduleName + "_descriptor_";
+  // The module name can be manipulated. At this point we need the original
+  // name.
+  auto originalModuleName =
```

Comment:
Is it possible to flip the logic here such that the module name is not changed but in the place it is printed the sanitized one is used instead? Needing to modify the IR and then add this additional untyped attribute (and the extra logic for whether or not to apply it resulting in the IR differing between modules) has a higher chance of breaking/introducing unexpected behavior than if every place you printed the module name you either passed in a string or did a `sanitizeSymbolName(moduleOp.getName())`. It puts the logic of "I'm emitting a C identifier and need to have it be valid" on the C emitter and not something inherent to the input IR, making it much less difficult to decide which name to use (sanitized/unsanitized) and much less harder to get phase ordering issues (hand-authored IR after the comvert to emit c but didn't sanitize, but there's the implicit assumption all names have been sanitized).

Reply:
@marbre and me just had a discussion about this. While we see your point we fear that we run into naming problems/clashes which will be extremely hard to debug. Also, our emitters (this one and the upstreamed one) would continue to diverge. We briefly discussed adding a sanitation pass, which - as the name states - only handles the sanitation, as a third option. What do you think?

Reply:
I too fear naming problems - we may just be seeing it from different angles :)
There's two approaches: one is to require all users of emitc users to consistently mangle names according to the emitc restrictions in user code, and the other is for emitc (a pass or pipeline or utility etc) to take care of it as an implementation detail. As a user I'm strongly in favor of the latter, and as maintainers you really don't want to have to go make these changes for every codebase that uses emitc and should be in favor of it too :)

Taking another look at it, this implementation is actually introducing naming conflicts: `@foo_bar` and `@foo<bar`. When changing symbol names you need to use the symbol table to unique them for you. And the moment someone goes and makes a `@foo$bar` it'll fail (or `@foo•bar`, etc). Doing this fixup is tricky, and it's definitely a battery that should be included in the box with emitc.

Strawman: upstream `mlir::emitc::sanitizeSymbol(StringRef symbol)` and `mlir::emitc::sanitizeSymbolOps(Region &region)`. For all the top-level symbol ops inside any operation all the fixup is nicely hidden (and using the proper SymbolTable tracking and faster symbol name replacement). As a user I wouldn't mind calling those two methods, and as things get weirder (think nested modules, or multiple modules linked together within a parent module, etc) the same methods just work: I'd be able to build a pass that did the sanitization while also making sure to preserve metadata that may be required, do my nesting, etc - vs a pass upstream that assumed a specific module structure.